### PR TITLE
POLIO-567 vaccine tab

### PR DESCRIPTION
**Pull request by quang-le**
Add a Vaccine Management tab to the main polio dashboard

## Self proof reading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented -  I shall hope so
- [X] Are my typescript files well typed 
- [X] Did I add translations 
- [X] My migrations file are included
- [ ] Are there enough tests - Nope :(
### Feature
- The screen is divided in 4 sections: Vaccines, Shipments, Form A and Destruction
- In the vaccines section:
    - The user can add vaccines for a total up to 3 (1 of each type)
    - The user can delete the vaccines one by one from the bottom up
    - The UI will show empty vaccine fields when deleting the last vaccine
    - When the user fills out a vaccine field, the values for reporting delays automatically fill up with default values if they don't have values yet
- In the shipment sections:
    - Same general behavior as vaccines except: 
        - Deleting the last remaining row will remove it from the UI
        - There's no limit to the number of shipment that can be added
        - Filling one field will prompt the validation to force the user to fill all shipment fields or leave them all blank
- Form A and Destruction sections: 
    - If one field is filled up, all related fields but also be filled

### Code
- Front-end
    - Added new components: `VaccineManagementForm`, `RoundVaccineForms` ,`RoundVaccineForm`, `ShipmentForms`, `ShipmentForm` 
    - Updated `useFormValidator` with new fields and new custom validation methods to force user to fill in related fields
- Back-end:
    - Added 2 models: `RoundVaccine` and `Shipment` with a ForeignKey to `Round`
    - Added `create` and `update` methods to `RoundSerializer` to handle the `vaccines` (RoundVaccine) and `shipments`(Shipment) fields

## How to test

- Go to a campaign
- Open the Vaccine Management Tab
- Test for the behavior described above under Feature

## Print screen / video

https://user-images.githubusercontent.com/38907762/191803835-ebba0380-bd2a-48f9-8580-e8aba9a3d56c.mov

## Notes

@olethanh Please have a good look at the `update`method in `RoundSerializer`

### CODE REVIEW

File: plugins/polio/js/src/forms/RoundVaccineForm.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    roundIndex,
+    vaccineIndex,
+    vaccineOptions,
+}) => {
+    const { formatMessage } = useSafeIntl();
+    const classes: Record<string, string> = useStyles();
+    const accessor = `rounds[${roundIndex}].vaccines[${vaccineIndex}]`;
+
+    return (
+        // <Grid container item xs={12} spacing={2}>
```
Comment (beygorghor) [inline plugins/polio/js/src/forms/RoundVaccineForm.tsx]: you can remove this

File: plugins/polio/js/src/forms/RoundVaccinesForm.tsx [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+                            key={`vaccine${index}`}
+                            vaccineIndex={index}
+                            roundIndex={roundIndex}
+                            vaccineOptions={polioVaccines}
+                        />
+                    );
+                })}
+            {vaccines.length === 0 && (
+                <RoundVaccineForm
+                    // eslint-disable-next-line react/no-array-index-key
```
Comment (beygorghor) [inline plugins/polio/js/src/forms/RoundVaccinesForm.tsx]: Key is not required here i hink

File: plugins/polio/js/src/forms/VaccineManangementForm.tsx [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                <Divider style={{ width: '100%' }} />
+                <Grid item xs={12}>
+                    <Box mt={1} mb={1}>
+                        <Typography variant="button">
+                            {formatMessage(MESSAGES.formA)}
+                        </Typography>
+                    </Box>
+                </Grid>
+                <Grid container direction="row" item xs={12} spacing={2}>
+                    <Grid item xs={3}>
```
Comment (beygorghor) [inline plugins/polio/js/src/forms/VaccineManangementForm.tsx]: Maybe use a separate component for FormA fields

File: plugins/polio/js/src/hooks/useFormValidator.js [hunk 4]
```diff
@@ -88,9 +89,257 @@ yup.addMethod(
...
+    },
+);
+
+yup.addMethod(
+    yup.number,
+    'hasAllFormAFieldsNumber',
+    function hasAllFormAFieldsNumber(formatMessage) {
+        return this.test('hasAllFormAFieldsNumber', '', (value, context) => {
+            const { path, createError, parent } = context;
+            const { forma_reception } = parent;
```
Comment (beygorghor) [inline plugins/polio/js/src/hooks/useFormValidator.js]: `const { forma_reception, forma_missing_vials,  forma_usable_vials, forma_unusable_vials} = parent;` ?

File: plugins/polio/js/src/hooks/useFormValidator.js [hunk 5]
```diff
@@ -88,9 +89,257 @@ yup.addMethod(
...
+        });
+    },
+);
+yup.addMethod(
+    yup.date,
+    'hasAllFormAFieldsDate',
+    function hasAllFormAFieldsDate(formatMessage) {
+        return this.test('hasAllFormAFieldsDate', '', (value, context) => {
+            const { path, createError, parent } = context;
+            const { forma_reception } = parent;
```
Comment (beygorghor) [inline plugins/polio/js/src/hooks/useFormValidator.js]: same here

File: plugins/polio/js/src/hooks/useFormValidator.js [hunk 6]
```diff
@@ -88,9 +89,257 @@ yup.addMethod(
...
+    },
+);
+yup.addMethod(
+    yup.date,
+    'hasAllShipmentFieldsDate',
+    function hasAllShipmentFieldsDate(formatMessage) {
+        return this.test('hasAllShipmentFieldsDate', '', (value, context) => {
+            const { path, createError, parent } = context;
+            const { date_reception } = parent;
+            const { estimated_arrival_date } = parent;
```
Comment (beygorghor) [inline plugins/polio/js/src/hooks/useFormValidator.js]: same

File: plugins/polio/js/src/hooks/useFormValidator.js [hunk 7]
```diff
@@ -88,9 +89,257 @@ yup.addMethod(
...
+    },
+);
+yup.addMethod(
+    yup.number,
+    'hasAllShipmentFieldsNumber',
+    function hasAllShipmentFieldsNumber(formatMessage) {
+        return this.test('hasAllShipmentFieldsNumber', '', (value, context) => {
+            const { path, createError, parent } = context;
+            const { date_reception } = parent;
+            const { estimated_arrival_date } = parent;
```
Comment (beygorghor) [inline plugins/polio/js/src/hooks/useFormValidator.js]: ...

File: plugins/polio/js/src/hooks/useFormValidator.js [hunk 8]
```diff
@@ -88,9 +89,257 @@ yup.addMethod(
...
+    },
+);
+yup.addMethod(
+    yup.string,
+    'hasAllShipmentFieldsString',
+    function hasAllShipmentFieldsString(formatMessage) {
+        return this.test('hasAllShipmentFieldsString', '', (value, context) => {
+            const { path, createError, parent } = context;
+            const { date_reception } = parent;
+            const { estimated_arrival_date } = parent;
```
Comment (beygorghor) [inline plugins/polio/js/src/hooks/useFormValidator.js]: ...

File: plugins/polio/js/src/pages/Dashboard.js [hunk 9]
```diff
@@ -190,7 +190,7 @@ const Dashboard = ({ router }) => {
                     const allRounds = (
                         <>
                             {row.rounds.map(r => (
-                                <li>
+                                <li key={`${r.number}-${r.started_at}`}>
```
Comment (beygorghor) [inline plugins/polio/js/src/pages/Dashboard.js:193]: This is also fixed on my PR [here](https://github.com/BLSQ/iaso/pull/17)

**Review by beygorghor — APPROVED** [review]
We can improve the view for smaller screens:
Screenshot 2022-09-23 at 11 54 34
Also some fields like on print screen pass to error state while only clicking on it.

Nothing blocking for me

File: plugins/polio/serializers.py [hunk 10]
```diff
@@ -240,12 +242,91 @@ class Meta:
...
 
     scopes = RoundScopeSerializer(many=True, required=False)
+    vaccines = RoundVaccineSerializer(many=True, required=False)
+    shipments = ShipmentSerializer(many=True, required=False)
+
+    @atomic
+    def create(self, validated_data):
+        vaccines = validated_data.pop("vaccines", [])
+        shipments = validated_data.pop("shipments", [])
+        print("VALIDATED DATA", validated_data)
```
Comment (olethanh) [inline plugins/polio/serializers.py:278]: print

File: plugins/polio/js/src/pages/Dashboard.js (hunk 9 above)
Comment (quang-le) [reply plugins/polio/js/src/pages/Dashboard.js:193]: next week ;-)

File: plugins/polio/js/src/forms/VaccineManangementForm.tsx (hunk 3 above)
Comment (quang-le) [reply plugins/polio/js/src/forms/VaccineManangementForm.tsx]: True. I'll make a second pass when updating the number input

**Comment by quang-le** [comment]
> We can improve the view for smaller screens: Screenshot 2022-09-23 at 11 54 34 Also some fields like on print screen pass to error state while only clicking on it.
> 
> Nothing blocking for me

I think the thing with the error state comes from the fact that for some parts of the form, all related fields must be filled in together. In your printscreen, you've filled out the vaccine for a shipment, which put all other fields for that shipment in error. When you clicked, `touched` switched to true, so the error was displayed. I can't really think of a workaround at the moment, though it's close to the behaviour we have with rounds

File: plugins/polio/serializers.py [hunk 11]
```diff
@@ -240,12 +242,91 @@ class Meta:
...
+                if shipment.round != instance:
+                    raise serializers.ValidationError({"shipments": "shipment is attached to wrong round"})
+            else:
+                shipment = Shipment.objects.create()
+            shipment_serializer = ShipmentSerializer(instance=shipment, data=shipment_data)
+            shipment_serializer.is_valid(raise_exception=True)
+            shipment_instance = shipment_serializer.save()
+            shipment_instances.append(shipment_instance)
+        # remove deleted shipments, ie existing shipments whose id wan't sent in the request
+        all_current_shipments = instance.shipments.all()
```
Comment (olethanh) [inline plugins/polio/serializers.py:322]: we delete the shipment but not the vaccine?

**Review by olethanh — APPROVED** [review]
Removed a debug print
It's not very clear when we start filling a shipement line that the Date need to be filled because they never get in red.
It should delete the vaccine line not linked to a round like we does for shipement
And fix the Log modification
But nothing blocking we can do it later

**Pull request merged**